### PR TITLE
Parse module elements in a block context

### PR DESCRIPTION
**Pull request by JsonFreeman**
We recently allowed type declarations in block contexts, but not module elements (modules, imports, exports, etc). This change adds parsing for module elements in block contexts, and produces the appropriate errors in the checker. This work has 3 benefits:

1. The user gets a better error, and error recovery in the parser is improved.
2. If we end up allowing module elements in a block context later, we would need to do this work.
3. It enables us to reduce/simplify some of the lookahead in the parser when determining whether something is a statement or module element (I have not done any lookahead removal).

### CODE REVIEW

File: src/compiler/binder.ts [hunk 1]
```diff
@@ -661,7 +661,11 @@ module ts {
         }
 
         function bindExportAssignment(node: ExportAssignment) {
-            if (node.expression.kind === SyntaxKind.Identifier) {
+            if (!container.symbol || !container.symbol.exports) {
+                // Export assignment in some sort of block construct
+                bindAnonymousDeclaration(node, SymbolFlags.Alias, "");
```
Comment (CyrusNajmabadi) [inline src/compiler/binder.ts]: can you use a dunder name instead?  I'm not very comfortable with entities having empty names.

**Comment by DanielRosenwasser** [comment]
Can you have a test for completion lists in the following?

```TypeScript
function f() {
    namespace n {
        interface I {
            x: number
        }
        /*1*/
    }
    /*2*/
}
/*3*/
```

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -11303,6 +11323,15 @@ module ts {
             }
         }
 
+        function checkGrammarModuleElementContext(node: ModuleElement, errorMessage: DiagnosticMessage): boolean {
+            if (node.parent.kind !== SyntaxKind.SourceFile && node.parent.kind !== SyntaxKind.ModuleBlock && node.parent.kind !== SyntaxKind.ModuleDeclaration) {
+                grammarErrorOnFirstToken(node, errorMessage);
```
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: can you unify the meaning of true/false in the grammar checks?

File: src/compiler/types.ts [hunk 3]
```diff
@@ -808,9 +808,7 @@ module ts {
         expression: UnaryExpression;
     }
 
-    export interface Statement extends Node, ModuleElement {
-        _statementBrand: any;
-    }
+    export interface Statement extends Node, ModuleElement { }
```
Comment (CyrusNajmabadi) [inline src/compiler/types.ts]: why get rid of _statementBrand?

**Comment by CyrusNajmabadi** [comment]
Update isReusableStatement to be the list of statements.

File: src/compiler/types.ts (hunk 3 above)
Comment (JsonFreeman) [inline src/compiler/types.ts]: I can remove ModuleElement

**Comment by JsonFreeman** [comment]
Will do, and I can remove the module element one.

File: src/compiler/binder.ts (hunk 1 above)
Comment (JsonFreeman) [inline src/compiler/binder.ts]: Fixed

**Comment by JsonFreeman** [comment]
@CyrusNajmabadi, should I remove ParsingContext.ModuleElements as well?

File: src/compiler/checker.ts (hunk 2 above)
Comment (JsonFreeman) [inline src/compiler/checker.ts]: I'll just switch this to fit with the original model where true means there was an error.

**Comment by JsonFreeman** [comment]
I've updated this pull request to address the feedback. Let me know if there is anything else that needs to be addressed.

File: src/compiler/parser.ts [hunk 4]
```diff
@@ -1814,7 +1789,7 @@ module ts {
             // the code would be implicitly: "name.keyword; identifierNameOrKeyword".
             // In the first case though, ASI will not take effect because there is not a
             // line terminator after the keyword.
-            if (scanner.hasPrecedingLineBreak() && scanner.isReservedWord()) {
+            if (scanner.hasPrecedingLineBreak() && isIdentifierOrKeyword()) {
```
Comment (CyrusNajmabadi) [inline src/compiler/parser.ts]: update the comment.

Comment (JsonFreeman) [inline src/compiler/parser.ts]: Fixed.

File: tests/cases/fourslash/completionWithDotFollowedByNamespaceKeyword.ts [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+////namespace A {
+////    function foo() {
+////        if (true) {
+////            B./**/
+////        namespace B {
+////            export function baz() { }
+////}
+
+goTo.marker();
+verify.completionListContains("baz", "function B.baz(): void");
```
Comment (CyrusNajmabadi) [inline tests/cases/fourslash/completionWithDotFollowedByNamespaceKeyword.ts]: nice.

**Pull request merged**